File: java/L000/LeetCode0057.java
package L000;

import java.util.ArrayList;
import java.util.List;

public class LeetCode0057 {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        List<int[]> res = new ArrayList<>();
        int i = 0;
        for (;i < intervals.length;++i) {
            if (newInterval[0] > intervals[i][1]) {
                int[] cur = new int[2];
                cur[0] = intervals[i][0];
                cur[1] = intervals[i][1];
                res.add(cur);
            } else {
                break;
            }
        }
        if (i == intervals.length || newInterval[1] < intervals[i][0]) {
            res.add(newInterval);
        } else {
            int[] cur = new int[2];
            cur[0] = Math.min(intervals[i][0], newInterval[0]);
            cur[1] = Math.max(intervals[i][1], newInterval[1]);
            for (i = i + 1;i < intervals.length;++i) {
                if (cur[1] >= intervals[i][0]) {
                    cur[1] = Math.max(cur[1], intervals[i][1]);
                } else {
                    break;
                }
            }
            res.add(cur);
        }
        for (;i < intervals.length;++i) {
            res.add(intervals[i]);
        }
        int[][] resArr = new int[res.size()][2];
        for (int k = 0;k < resArr.length;++k) {
            resArr[k] = res.get(k);
        }
        return resArr;
    }
}

File: cpp/L000/LeetCode0057.cc
#include <algorithm>
#include <vector>
using std::vector;

class Solution {
public:
    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
        vector<vector<int>> res;
        std::size_t n{intervals.size()};
        std::size_t i{0};
        while (i < n && intervals[i][1] < newInterval[0]) {
            res.push_back(intervals[i]);
            ++i;
        }
        if (i == n) {
            res.push_back(newInterval);
            return res;
        }
        if (newInterval[1] < intervals[i][0]) {
            res.push_back(newInterval);
        } else {
            int left{std::min(intervals[i][0], newInterval[0])}, right{std::max(newInterval[1], intervals[i][1])};
            ++i;
            while (i < n && intervals[i][0] <= right) {
                right = std::max(right, intervals[i][1]);
                ++i;
            }
            res.push_back({left, right});
        }
        while (i < n) {
            res.push_back(intervals[i]);
            ++i;
        }
        return res;
    }
};